src/utils/supergraph_config.rs

```diff
@@ -0,0 +1,1169 @@
+use std::str::FromStr;
+
+use anyhow::anyhow;
+use apollo_federation_types::build::{BuildError, BuildErrors, SubgraphDefinition};
+use apollo_federation_types::config::{
+    FederationVersion, SchemaSource, SubgraphConfig, SupergraphConfig,
+};
+use apollo_parser::{cst, Parser};
+use rayon::iter::{IntoParallelIterator, ParallelIterator};
+
+use rover_client::blocking::{GraphQLClient, StudioClient};
+use rover_client::operations::subgraph;
+use rover_client::operations::subgraph::fetch::SubgraphFetchInput;
+use rover_client::operations::subgraph::fetch_all::SubgraphFetchAllInput;
+use rover_client::operations::subgraph::introspect::SubgraphIntrospectInput;
+use rover_client::operations::subgraph::{fetch, introspect};
+use rover_client::shared::GraphRef;
+use rover_client::RoverClientError;
+use rover_std::{Fs, Style};
+
+use crate::options::ProfileOpt;
+use crate::utils::client::StudioClientConfig;
+use crate::utils::expansion::expand;
+use crate::utils::parsers::FileDescriptorType;
+use crate::{RoverError, RoverErrorSuggestion, RoverResult};
+
+/// Nominal type that captures the behavior of collecting remote subgraphs into a
+/// [`SupergraphConfig`] representation
+#[derive(Clone, Debug)]
+pub struct RemoteSubgraphs(SupergraphConfig);
+
+impl RemoteSubgraphs {
+    /// Fetches [`RemoteSubgraphs`] from Studio
+    pub fn fetch(
+        client: &StudioClient,
+        federation_version: &FederationVersion,
+        graph_ref: &GraphRef,
+    ) -> RoverResult<RemoteSubgraphs> {
+        let subgraphs = subgraph::fetch_all::run(
+            SubgraphFetchAllInput {
+                graph_ref: graph_ref.clone(),
+            },
+            client,
+        )?;
+        let subgraphs = subgraphs
+            .iter()
+            .map(|subgraph| {
+                (
+                    subgraph.name().clone(),
+                    SubgraphConfig {
+                        routing_url: subgraph.url().clone(),
+                        schema: SchemaSource::Sdl {
+                            sdl: subgraph.sdl().clone(),
```

Comment:
are sdls pretty chonky? wondering if we absolutely need to clone them or if there's another way to get them where we need them without having to go down that route (mostly worried about a crazy amount of strings being processed when the number of subgraphs is non-trivial); might be an empirical question of what the actual performance looks like when we run this against larger sets of subgraphs and may end up not mattering, but worth worrying about so we don't slow down the `supergraph compose` command

Reply:
This is a great callout, and I think we can be a bit clever here. Because we're consuming the subgraphs vector we can use `into_iter` and then if we implement `From<Subgraph> for SubgraphConfig` we can properly consume the subgraphs rather than cloning, so we should only ever have 1 string around for the SDL if it is chonky. It means we have to clone the name otherwise we've got references that are a bit of a mess but even with hundreds of subgraphs the overhead would be bytes which doesn't bother me.